DllImportGenerator/DllImportGenerator/ManualTypeMarshallingAnalyzer.cs

```diff
@@ -0,0 +1,342 @@
+                Category,
+                DiagnosticSeverity.Warning,
+                isEnabledByDefault: true,
+                description: "A type that supports marshalling from managed to native by pinning should also support marshalling from managed to native where pinning is impossible.");
+
+        public readonly static DiagnosticDescriptor StackallocMarshallingShouldSupportAllocatingMarshallingFallbackRule =
+            new DiagnosticDescriptor(
+                "INTEROPGEN011",
+                "StackallocMarshallingShouldSupportAllocatingMarshallingFallback",
+                "Native type '{0}' has a stack-allocating constructor does not support marshalling in scenarios where stack allocation is impossible.",
+                Category,
+                DiagnosticSeverity.Warning,
+                isEnabledByDefault: true,
+                description: "A type that supports marshalling from managed to native by stack allocation should also support marshalling from managed to native where stack allocation is impossible.");
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => 
+            ImmutableArray.Create(
+                BlittableTypeMustBeBlittableRule,
+                CannotHaveMultipleMarshallingAttributesRule,
+                NativeTypeMustBeNonNullRule,
+                NativeTypeMustBeBlittableRule,
+                GetPinnableReferenceReturnTypeBlittableRule,
+                NativeTypeMustBePointerSizedRule,
+                NativeTypeMustHaveRequiredShapeRule,
+                ValuePropertyMustHaveSetterRule,
+                ValuePropertyMustHaveGetterRule,
+                GetPinnableReferenceShouldSupportAllocatingMarshallingFallbackRule,
+                StackallocMarshallingShouldSupportAllocatingMarshallingFallbackRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterSymbolAction(AnalyzeTypeDefinition, SymbolKind.NamedType);
+            context.RegisterSymbolAction(AnalyzeElement, SymbolKind.Parameter, SymbolKind.Field);
+            context.RegisterSymbolAction(AnalyzeReturnType, SymbolKind.Method);
+        }
+
+        private void AnalyzeTypeDefinition(SymbolAnalysisContext context)
+        {
+            INamedTypeSymbol type = (INamedTypeSymbol)context.Symbol;
+            var generatedMarshallingAttribute = context.Compilation.GetTypeByMetadataName("System.Runtime.InteropServices.GeneratedMarshallingAttribute");
+            var blittableTypeAttribute = context.Compilation.GetTypeByMetadataName("System.Runtime.InteropServices.BlittableTypeAttribute");
+            var nativeMarshallingAttribute = context.Compilation.GetTypeByMetadataName("System.Runtime.InteropServices.NativeMarshallingAttribute");
```

Comment:
Actually, I think we should avoid registering for the symbol actions at all if these types don't exist - i.e. use `RegisterCompilationStartAction`, try to get these known types in that action and only register the symbol actions if those types exist (and reuse them instead of looking them up every time).

Reply:
I wasn't sure which action allowed us to pull the symbols early. I'll add that and cache the type lookup results!